### DISCUSSION TITLE
Freezing InlineAutoData

### DISCUSSION THREAD

**Discussion by isberg** [AutoFixture/AutoFixture#1308] [Show and tell]
A while ago I was surprised that I could not get AutoFixture to Freeze my InlineAutoData:

``` csharp
    [Theory,
        InlineAutoData(),
        InlineAutoData(42)
    ]
    void InlineAutoData_ShouldBeFreezable([Frozen]int secret, Vault sut)
    {
        Assert.True(secret == sut.Secret);
    }
    
    record Vault
    {
        public Vault(int secret) => Secret = secret;
        public int Secret { get; }
    }
```

After learning plenty about AutoFixture I understood why it does not work (and why it is impossible?). However a while ago I decided to find out exactly why it was impossible and this is what I came up with:

```csharp
    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
    public class InlineAutoDataAttribute : AutoFixture.Xunit2.InlineAutoDataAttribute
    {
        private IFixture fixture => ((AutoDataAttribute)base.AutoDataAttribute).Fixture;
    
        public InlineAutoDataAttribute(params object[] values)
            : base(new AutoDataAttribute(), values)
        {
        }
    
        public override IEnumerable<object[]> GetData(MethodInfo testMethod)
        {
            foreach (var (parameter, value) in testMethod.GetParameters().Zip(Values))
            {
                if (parameter.CustomAttributes.Any(x => x.AttributeType == typeof(FrozenAttribute)))
                {
                    fixture
                        .Customizations
                        .Insert(0, new FilteringSpecimenBuilder(
                            new FixedBuilder(value),
                            new ExactTypeSpecification(parameter.ParameterType)));
                }
            }
    
            return base.GetData(testMethod);
        }
    }
```

This actually seems to work (for AutoFixture.Xunit 4.17.0), but it suffers from using the obsolete Fixture property of AutoDataAttribute. I would be happy to know if there is a better way to do this or if might suffers from some other weaknesses than being obsolete. :-)

**Comment by isberg** [comment]
Or maybe it is better like this, avoiding using obsolete properties:

``` csharp
    [DataDiscoverer("AutoFixture.Xunit2.NoPreDiscoveryDataDiscoverer", "AutoFixture.Xunit2")]
    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
    public class InlineFreezingAutoDataAttribute : InlineAutoDataAttribute
    {
        internal DynamicSpecimenBuilder SpecimenBuilder;
    
        public InlineFreezingAutoDataAttribute(params object[] values)
            : this(new DynamicSpecimenBuilder(), values)
        {
        }
    
        private InlineFreezingAutoDataAttribute(DynamicSpecimenBuilder specimenBuilder, params object[] values)
            : base(new CustomAutoDataAttribute(() => new Fixture().Customize(specimenBuilder.ToCustomization())), values)
        {
            SpecimenBuilder = specimenBuilder;
        }
    
        public override IEnumerable<object[]> GetData(MethodInfo testMethod)
        {
            foreach (var (parameter, value) in testMethod.GetParameters().Zip(Values))
            {
                if (parameter.CustomAttributes.Any(x => x.AttributeType == typeof(FrozenAttribute)))
                {
                    SpecimenBuilder
                        .Add(new FilteringSpecimenBuilder(
                            new FixedBuilder(value),
                            new ExactTypeSpecification(parameter.ParameterType)));
                }
            }
    
            return base.GetData(testMethod);
        }
    
        class CustomAutoDataAttribute : AutoDataAttribute
        {
            public CustomAutoDataAttribute(Func<IFixture> fixtureFactory)
                : base(fixtureFactory) { }
        }
    }
    
    public class DynamicSpecimenBuilder : ISpecimenBuilder
    {
        public List<ISpecimenBuilder> Builders { get; } = new ();
    
        public object Create(object request, ISpecimenContext context)
            => Builders
                .Select(x => x.Create(request, context))
                .FirstOrDefault(x => !(x is NoSpecimen), new NoSpecimen());
    
        public void Add(ISpecimenBuilder builder) => Builders.Add(builder);
    }
```

**Reply by aivascu** [reply]
The AutoData attributes will receive a rewrite in both xUnit and NUnit libaries, in the next version. The intention is to design these attributes so they could support the Freeze attribute as an inject as well as other features like ranges and permutations from NUnit. The solution to injected test parameters is indeed to create specimen that intercept requests for specific types. The difference from the original `Frozen` attribute is that the value will not be resolved from the fixture but rather from the test case provider.

**Reply by isberg** [reply]
@aivascu any guess when the ability to use Frozen with InlineAutoData will make it into the preview of version 5? :-)

**Reply by aivascu** [reply]
@isberg now that the `netstandard2.1` integration is out of the way the next big item on my todo list is the AutoData attributes (inline, member and class). These should include the `Frozen` attribute working out of the box with any parameters, either provided by autofixture or by the test case source. In terms of time, I'd say maybe a week or two depending on my availability.

**Reply by Kralizek** [reply]
@aivascu will you also look at nunit attributes?

**Reply by aivascu** [reply]
@Kralizek yes that's the plan, however I think I'll start with xunit since that has an easier internal API.

**Reply by isberg** [reply]
@aivascu is there any way to be of help regarding the new AutoData + Frozen + Inline implementation? Is there any forum where this is already discussed? :-)

**Reply by aivascu** [reply]
@isberg sorry for the late reply, lately I have very little spare time. The discussions around the Frozen attribute is spread between multiple threads in the repository, I could try to find them but I don't think they are relevant at this point.
The best source of information on the topic is [this](https://github.com/aivascu/AutoFixture/tree/dataattributes) branch where I was working on the overhaul of the autodata attributes for now mainly XUnit2.
The freezing of inline values should theoretically work however from what I remember I haven't even finished the unit test coverage. If you wish to help you are welcome to review it and provide your feedback.
I hope to return to active development after July, however for now I cannot make any promises. My work schedule has been crazy since late January.

**Reply by isberg** [reply]
Thanks @aivascu! I will check out the branch. I kind of expected this was the reason and situation.
